doc/api/domain.markdown

```diff
@@ -2,304 +2,265 @@
+are intentionally written to randomly throw errors. When these occur, the
+worker process shuts down immediately, leading to an ungraceful exit of the
+master process:
+
+    // An example of what not to do
+    const cluster = require('cluster');
+    
+    if (cluster.isMaster) {
+      cluster.fork();
+      cluster.fork();
+    } else {
+      function randomError() {
+        var r = Math.floor(Math.random() * 11);
+        if (r >= 7) throw new Error();
+        setTimeout(randomError, 1000);
       }
-    });
+      setTimeout(randomError, 1000);
+    }
 
-    // Because req and res were created before this domain existed,
-    // we need to explicitly add them.
-    // See the explanation of implicit vs explicit binding below.
-    d.add(req);
-    d.add(res);
+The `domain` module can be introduced within the worker process to provide
+graceful error handling as in the following modified version. However, since
+the errors are simply being ignored, resources are being leaked and held on
+to.
+
+    // Another example of what not to do
+    const cluster = require('cluster');
+
+    if (cluster.isMaster) {
+      cluster.fork();
+      cluster.fork();
+    } else {
+      const domain = require('domain').create();
+      domain.on('error', (err) => {
+        console.log('There was an error in the worker');
+      });
+      domain.run(() => {
+        function randomError() {
+          var r = Math.floor(Math.random() * 11);
+          if (r >= 7) throw new Error();
+          setTimeout(randomError, 1000);
+        }
+        setTimeout(randomError, 1000);
+      });
+    }
 
-    // Now run the handler function in the domain.
-    d.run(() => {
-      handleRequest(req, res);
-    });
-  });
-  server.listen(PORT);
-}
-
-// This part isn't important.  Just an example routing thing.
-// You'd put your fancy application logic here.
-function handleRequest(req, res) {
-  switch(req.url) {
-    case '/error':
-      // We do some async stuff, and then...
-      setTimeout(() => {
-        // Whoops!
-        flerb.bark();
+The better approach is to handle the error, disconnect the worker explicitly
+from the master process, allow the failing worker process to exit, then have
+the master create a new worker to take it's place:
+
+    const cluster = require('cluster');
+
+    if (cluster.isMaster) {
+      cluster.fork();
+      cluster.fork();
+    
+      // Spawn a new worker when one disconnects   
+      cluster.on('disconnect', (worker) => {
+        console.error(`${worker.id} failed. spawning a replacement.`);
+        cluster.fork();
+      });
+      
+    } else {
+      const domain = require('domain').create();
+      domain.on('error', (err) => {
+        console.log(`There was an error in ${cluster.worker.id}.`);
+        
+        try {
+          // Make sure the worker closes down within 30 seconds.
+          // But don't keep the process open just for that!
```

Comment:
We should make clear that it's the call to `unref` that prevents the timeout from keeping the loop open. Readers of this code may not be familiar with that aspect of the libuv event loop and timers.